core/utils/metering.h

```diff
@@ -0,0 +1,247 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef BESS_UTILS_METERING_H_
+#define BESS_UTILS_METERING_H_
+
+#include <string>
+#include <type_traits>
+#include <vector>
+
+#include "../message.h"
+#include "../metadata.h"
+#include "../module.h"
+#include "../packet.h"
+#include "bits.h"
+#include "cuckoo_map.h"
+#include "endian.h"
+#include "format.h"
+#include <rte_config.h>
+#include <rte_hash_crc.h>
+#include <rte_meter.h>
+
+#define MAX_FIELDS 8
+#define MAX_FIELD_SIZE 8
+
+static_assert(MAX_FIELD_SIZE <= sizeof(uint64_t),
+              "field cannot be larger than 8 bytes");
+
+#define HASH_KEY_SIZE (MAX_FIELDS * MAX_FIELD_SIZE)
+
+#if __BYTE_ORDER__ != __ORDER_LITTLE_ENDIAN__
+#error this code assumes little endian architecture (x86)
+#endif
+
+struct MeteringField {
+  int attr_id;
+  int offset;
+  int pos;
+  int size;
+};
+
+namespace bess {
+namespace utils {
+
+using Error = std::pair<int, std::string>;
+
+struct MeteringKey {
+  uint64_t u64_arr[MAX_FIELDS];
+} __attribute__((packed));
+
+// Equality operator for two MeteringKeys
+class MeteringKeyEq {
+ public:
+  explicit MeteringKeyEq(size_t len) : len_(len) {}
+
+  bool operator()(const MeteringKey &lhs, const MeteringKey &rhs) const {
+    promise(len_ >= sizeof(uint64_t));
+    promise(len_ <= sizeof(MeteringKey));
+
+    for (size_t i = 0; i < len_ / 8; i++) {
+      if (lhs.u64_arr[i] != rhs.u64_arr[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+ private:
+  size_t len_;
+};
+
+// Hash function for an MeteringKey
+class MeteringKeyHash {
+ public:
+  explicit MeteringKeyHash(size_t len) : len_(len) {}
+
+  HashResult operator()(const MeteringKey &key) const {
+    HashResult init_val = 0;
+
+    promise(len_ >= sizeof(uint64_t));
+    promise(len_ <= sizeof(MeteringKey));
+
+#if __x86_64
+    for (size_t i = 0; i < len_ / 8; i++) {
+      init_val = crc32c_sse42_u64(key.u64_arr[i], init_val);
+    }
+    return init_val;
+#else
+    return rte_hash_crc(&key, len_, init_val);
+#endif
+  }
+
+ private:
+  size_t len_;
+};
+
+template <typename T>
+class Metering {
+ public:
+  struct rte_hash_parameters dpdk_params {
+    .name = "Metering", .entries = 1 << 20, .reserved = 0,
```

Comment:
open a issue to track removing 1M entry hardcode across all 3 modules

Reply:
we can also think if control path shall provide that option , would be useful to track how many entries has been and for 1SMF:Nupf scenarios, if we are thinking to do in future.  Hardcoding through config file is an other option. Pls go ahead, and implement it after discussion.